src/main/java/com/google/devtools/build/lib/analysis/ConfiguredTargetFactory.java

```diff
@@ -304,6 +308,76 @@ private ConfiguredTarget createRule(
                     prerequisiteMap.values()))
             .build();
 
+    if (!"toolchain".equals(rule.getRuleClass())) {
+      boolean isIncompatible = false;
+
+      // This is incompatible if one of the dependencies is as well.
+      for (ConfiguredTargetAndData infoCollection : prerequisiteMap.values()) {
+        if (infoCollection.getConfiguredTarget().getProvider(IncompatiblePlatformProvider.class) != null) {
+          isIncompatible = true;
+        }
+      }
+
+      // This is incompatible if explicitly specified to be.
+      if (!isIncompatible && ruleContext.attributes().has("target_compatible_with")) {
+        Iterable<ConstraintValueInfo> constraintValues =
+            PlatformProviderUtils.constraintValues(
+                ruleContext.getPrerequisites("target_compatible_with", TransitionMode.DONT_CHECK));
+        for (ConstraintValueInfo constraintValue : constraintValues) {
+          if (!ruleContext.targetPlatformHasConstraint(constraintValue)) {
+            isIncompatible = true;
+          }
+        }
+      }
+      if (isIncompatible) {
+        NestedSetBuilder<Artifact> filesToBuildBuilder =
+            NestedSetBuilder.<Artifact>stableOrder();
+
+        ImmutableList<Artifact> outputArtifacts = ruleContext.getOutputArtifacts();
+
+        if (ruleContext.isTestTarget() && outputArtifacts.size() == 0) {
+          // TODO(philsc): Figure out how to avoid this. Right now test targets require the
+          // RunfilesSupport to be added to the RuleConfiguredTargetBuilder. Can we use an input for
+          // this? How do I get access to inputs?
+          outputArtifacts = ImmutableList.of(ruleContext.createOutputArtifact());
+        }
```

Comment:
This feels like a hack. I wasn't sure what the right answer here is.

If there are no output artifacts (e.g. `sh_test`) then how do I set the "executable" of this `ConfiguredTarget`? Ideally, I'd use one of the input files for this, but it wasn't obvious to me how to access them. I could probably iterate over `prerequisiteMap.values()` and see if there are any `InputFileConfiguredTarget` instances in the configured target.

Anyway, I'll think about this some more next week.

Reply:
I can't remember... are there really tests with no output artifacts? Like is it really possible for `outputArtifacts.size()` to be 0?

Reply:
This is what `sh_test` appears to do. I'm not really sure why since it produces a symlink to the shell script.

More investigation is required.

Reply:
Looks like the very first thing that `ShBinary.create()` does is to call `ruleContext.createOutputArtifact()`. Since we're not calling `ShBinary.create()` we have to create the output artifcact ourselves.

I'm a little more confident that _something_ along these lines is the right answer, but it still seems like a suboptimal answer.

Reply:
I spent time today trying to convince bazel that it's okay not to have a "main executable" for tests, but dug myself into a very deep hole.

Is there some kind of "dummy executable" rules can use? Like an `Artifact` equivalent of `FailAction`?

On that note, it looks like `cc_test` doesn't require this logic. I couldn't get `py_test` to work in my shell integration tests. So it's possible `sh_test` is indeed the only test that generates the executable in the factory's `create()` call. I'm assuming that Starlark code doesn't have this problem. But I should really verify this.

Reply:
I think `AnalysisTestActionBuilder.java` does something similar. I'm not seeing here where you actually put an implementing (like `exit 0` on Linux) into the test binary though.

(I have a pet project for Bazel itself to just bypass to execute that action via whatever means to avoid wasting test runner resources but that is a tangential discussion.)

Reply:
Interesting! Thanks @sdtwigg . I'll take a look at that.

I skipped implementation of the test binary because it appears not to be necessary. Maybe because the file never gets invoked? I've actually not worried about it, but maybe I should :P

Reply:
@sdtwigg , I looked at `AnalysisTestActionBuilder.java` and couldn't figure out a good way to make use of what it does. Thank you very much for the help!

I ended up documenting the hack where hopefully it's obvious enough that it'd be nice to solve it, but is currently non-trivial.